// webppl/src/aggregation/ScoreAggregator.js
'use strict';
var ad = require('../ad.js');
var assert = require('assert');
var _ = require('underscore');
var dists = require('../dists');
var util = require('../util');
function logsumexp(a, b) {
    assert.ok(ad.scalar.pneq(a, ad.scalar.sub(0, Infinity)) || ad.scalar.pneq(b, ad.scalar.sub(0, Infinity)));
    var m = ad.scalar.max(a, b);
    return ad.scalar.add(ad.scalar.log(ad.scalar.add(ad.scalar.exp(ad.scalar.sub(a, m)), ad.scalar.exp(ad.scalar.sub(b, m)))), m);
}
var ScoreAggregator = function () {
    this.dist = {};
};
Object.defineProperties(ScoreAggregator.prototype, {
    size: {
        get: function () {
            return _.size(this.dist);
        }
    }
});
ScoreAggregator.prototype.add = function (value, score) {
    if (ad.scalar.peq(score, ad.scalar.sub(0, Infinity))) {
        return;
    }
    var k = util.serialize(value);
    if (ad.scalar.peq(this.dist[k], undefined)) {
        this.dist[k] = {
            score: ad.scalar.sub(0, Infinity),
            val: value
        };
    }
    this.dist[k].score = logsumexp(this.dist[k].score, score);
};
function normalize(dist) {
    var logNorm = _.reduce(dist, function (acc, obj) {
        return logsumexp(acc, obj.score);
    }, ad.scalar.sub(0, Infinity));
    return _.mapObject(dist, function (obj) {
        return {
            val: obj.val,
            prob: ad.scalar.exp(ad.scalar.sub(obj.score, logNorm))
        };
    });
}
ScoreAggregator.prototype.toDist = function () {
    return new dists.Marginal({ dist: normalize(this.dist) });
};
module.exports = ScoreAggregator;
